api/queue/dispatcher.go

```diff
@@ -0,0 +1,131 @@
+package queue
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gojek/merlin/log"
+	"github.com/jinzhu/gorm"
+)
+
+type Producer interface {
+	EnqueueJob(job *Job) error
+}
+
+type Consumer interface {
+	RegisterJob(jobName string, jobFn func(*Job) error)
+	Start()
+	Stop()
+}
+
+type Dispatcher struct {
+	sync.Mutex
+	db         *gorm.DB
+	workers    []*worker
+	jobFuncMap map[string]JobFn
+	jobChan    chan *Job
+}
+
+type JobFn func(*Job) error
+
+type Config struct {
+	NumWorkers int
+	Db         *gorm.DB
+}
+
+type Status string
+
+type Argument map[string]interface{}
+
+type Job struct {
+	ID        int64     `json:"id" gorm:"primary_key;"`
+	Arguments Argument  `json:"arguments"`
+	Completed bool      `json:"completed"`
+	Name      string    `json:"name"`
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
+func (a Argument) Value() (driver.Value, error) {
+	return json.Marshal(a)
+}
+
+func (a *Argument) Scan(value interface{}) error {
+	b, ok := value.([]byte)
+	if !ok {
+		return errors.New("type assertion to []byte failed")
+	}
+
+	return json.Unmarshal(b, &a)
+}
+
+func NewDispatcher(cfg Config) *Dispatcher {
+	workers := make([]*worker, 0, cfg.NumWorkers)
+	jobChan := make(chan *Job)
+	jobFuncMap := make(map[string]JobFn)
+	for i := 0; i < cfg.NumWorkers; i++ {
+		worker := newWorker(cfg.Db, jobChan)
+		workers = append(workers, worker)
+	}
+	return &Dispatcher{
+		workers:    workers,
+		db:         cfg.Db,
+		jobFuncMap: jobFuncMap,
+		jobChan:    jobChan,
+	}
+}
+
+func (d *Dispatcher) EnqueueJob(job *Job) error {
+	var savedJob Job
+	if err := d.db.Save(job).Scan(&savedJob).Error; err != nil {
+		log.Errorf("Failed to save job %d with error: %v", job.ID, err)
+		return err
+	}
+	go func() {
```

Comment:
What is this code for?

Reply:
so for enqueuing job, queue will insert the job record to DB if it success it will send the job into channel, since the worker will consume job from values from channel